HCAL TP for Plan 1

matz-e opened the pull request:
Contains several previous PR, fixed:

* #17523, this PR supercedes that one.
* #17469 with enum `TriggerMode` meaning of the enumeration values unchanged.
* #17260, which got reverted with one of the L1T PR that was reverted.

@davidlange6 I fixed the RelVal 10826.0 by keeping the meaning of the trigger mode consistent, I had added the new 2016 and Plan 1 definition and changed the underlying values, which had already been saved in the geometry DB.

matz-e committed:
Increase robustness of HCAL TP LUT generation.

The current code crashes for Phase2.  Instead of generating all possible
DetIds and testing which ones are allowed, it is a lot cleaner to just
iterate over the produced metadata ids.

matz-e committed:
Filter out invalid ids.

matz-e committed:
Add more trigger geometry modes for Plan1 and Phase2.

matz-e committed:
Fix 2017 trigger mode to use QIE10 in HF.

matz-e committed:
Simplify branches for HT validity.

matz-e committed:
Fix branch.

matz-e committed:
Add Plan1 glue for HCAL TP.

matz-e committed:
Fix the values of the trigger modes to match the ones already in the DB.

cmsbuild commented:
A new Pull Request was created by  @matz-e (Matthias Wolf) for CMSSW_9_0_X.

It involves the following packages:

CalibCalorimetry/CaloTPG
CalibCalorimetry/HcalTPGAlgos
Geometry/CaloTopology
Geometry/HcalCommonData
Geometry/HcalTowerAlgo


@ghellwig, @civanch, @Dr15Jones, @arunhep, @ianna, @mdhildreth, @cmsbuild, @rekovic, @franzoni, @cerminar, @mmusich, @mulhearn, @davidlange6 can you please review it and eventually sign? Thanks.
@ghellwig, @tocheng this is something you requested to watch as well.
@davidlange6, @smuzaffar you are the release manager for this.

cms-bot commands are listed here cms-sw/cmssw#13028

civanch commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 9c7c69d4203cf4d932921a4d4103956b8dc7cfbd
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17540/17829/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17540/17829/summary.html

franzoni commented on CalibCalorimetry/CaloTPG/src/CaloTPGTranscoderULUT.cc:
```diff
@@ -204,12 +204,29 @@ CaloTPGTranscoderULUT::getOutputLUTSize(const HcalTrigTowerDetId& id) const
+            return QIE10_OUTPUT_LUT_SIZE;
+      case HcalTopologyMode::TriggerMode_2017plan1:
+         if (theTopology->dddConstants()->isPlan1(id))
+            return QIE11_OUTPUT_LUT_SIZE;
```
hello @matz-e 
detIdSp_ https://github.com/matz-e/cmssw/blob/9c7c69d4203cf4d932921a4d4103956b8dc7cfbd/Geometry/HcalCommonData/interface/HcalDDDRecConstants.h#L136   is empty by construction, for now
which in turns means isPlan1(...) always returns false.

matz-e replied:
Doesn't it get filled [here](https://github.com/cms-sw/cmssw/blob/a85956d574ec10a14b2dba8a1d503d4ddd0bb208/Geometry/HcalCommonData/src/HcalDDDRecConstants.cc#L896)?

matz-e commented:
I've noticed that this breaks various era-conditionals and will update with a fix later today.

matz-e committed:
Keep messy special statements at bay.

This makes the old 2017, now 2018 trigger mode value (2) a special "legacy"
item, and keeps the new trigger modes consistent above that in the enum.
Allows to keep comparisons in the code as long as the legacy value is
accounted for.

cmsbuild commented:
Pull request #17540 was updated. @ghellwig, @civanch, @Dr15Jones, @arunhep, @ianna, @mdhildreth, @cmsbuild, @rekovic, @franzoni, @cerminar, @mmusich, @mulhearn, @davidlange6 can you please check and sign again.

kpedro88 commented:
@matz-e ready to test?

matz-e commented:
Should be ready. I should note that Salavat's simple pgun config still fails for me. At least there should be no regression.

kpedro88 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: fe19e4f9c73517240219bec3d12401848bd00d50
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17540/17841/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17540/17841/summary.html

ianna commented:
+1

rekovic commented:
+1

franzoni commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_9_0_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @smuzaffar

davidlange6 commented on CalibCalorimetry/HcalTPGAlgos/src/HcaluLUTTPGCoder.cc:
```diff
@@ -238,126 +238,123 @@ void HcaluLUTTPGCoder::update(const HcalDbService& conditions) {
-    } // for iphi
-  }// for subdet
+  for (const auto& id: metadata->getAllChannels()) {
+     if (not (id.det() == DetId::Hcal and topo_->valid(id)))
```
is this if statement ever false in normal conditions? Eg, if its true is that a sign of corrupt data and/or incomplete conditions?

matz-e replied:
When I implemented this, there were IDs that were either invalid HCAL IDs or not allowed by the topology… I'm not sure anymore which IDs exactly are skipped by this.

ianna replied:
indeed there are such IDs. Verification of validity by topology supposed to be a temporary work-around, I think.

davidlange6 replied:
ok, thanks - is a dangerous condition to leave around, so lets try to clean up after than plan1 issues are behind us

rekovic replied:
@matz-e Was this not the fix when simHCALTriggerPrimitives in sequence for Phase2?

matz-e replied:
Yes, that was included in there, too. It somehow got reverted with the reversal of one of the L1T PRs, I think.

davidlange6 commented:
+1
